Sources/ConnectionPoolModule/PoolStateMachine.swift

```diff
@@ -84,10 +116,452 @@ struct PoolStateMachine<
+
+        case .shuttingDown:
+            return .none()
+
+        case .shutDown:
+            return .init(request: .none, connection: .none)
+        }
+    }
+
+    @inlinable
+    /*private*/ mutating func handleAvailableConnection(
+        index: Int,
+        availableContext: ConnectionGroup.AvailableConnectionContext
+    ) -> Action {
+        // this connection was busy before
+        let requests = self.requestQueue.pop(max: availableContext.info.availableStreams)
+        if !requests.isEmpty {
+            let leaseResult = self.connections.leaseConnection(at: index, streams: UInt16(requests.count))
+            return .init(
+                request: .leaseConnection(requests, leaseResult.connection),
+                connection: .cancelTimers(.init(leaseResult.timersToCancel))
+            )
+        }
+
+        switch availableContext.use {
+        case .persisted, .demand:
+            switch availableContext.info {
+            case .leased:
+                return .none()
+
+            case .idle:
+                let timers = self.connections.parkConnection(at: index).map(self.mapTimers)
+
+                return .init(
+                    request: .none,
+                    connection: .scheduleTimers(timers)
+                )
+            }
+
+        case .overflow:
+            let closeAction = self.connections.closeConnectionIfIdle(at: index)
+            return .init(
+                request: .none,
+                connection: .closeConnection(closeAction.connection, closeAction.timersToCancel)
+            )
+        }
+
+    }
+
+    @inlinable
+    /* private */ func mapTimers(_ connectionTimer: ConnectionTimer) -> Timer {
+        switch connectionTimer.usecase {
+        case .backoff:
+            return Timer(
+                connectionTimer,
+                duration: Self.calculateBackoff(failedAttempt: self.failedConsecutiveConnectionAttempts)
+            )
+
+        case .keepAlive:
+            return Timer(connectionTimer, duration: self.configuration.keepAliveDuration!)
+
+        case .idleTimeout:
+            return Timer(connectionTimer, duration: self.configuration.idleTimeoutDuration)
+
+        }
+    }
+}
+
+@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
+extension PoolStateMachine {
+    /// Calculates the delay for the next connection attempt after the given number of failed `attempts`.
+    ///
+    /// Our backoff formula is: 100ms * 1.25^(attempts - 1) that is capped of at 1 minute.
+    /// This means for:
+    ///   -  1 failed attempt :  100ms
+    ///   -  5 failed attempts: ~300ms
+    ///   - 10 failed attempts: ~930ms
+    ///   - 15 failed attempts: ~2.84s
+    ///   - 20 failed attempts: ~8.67s
+    ///   - 25 failed attempts: ~26s
+    ///   - 29 failed attempts: ~60s (max out)
+    ///
+    /// - Parameter attempts: number of failed attempts in a row
+    /// - Returns: time to wait until trying to establishing a new connection
+    @usableFromInline
```

Comment:
I'd recommend mentioning the jitter explicitly (rather than implying it with the `~` for "approximately")